addons/sale_project/models/sale_order.py

```diff
@@ -353,3 +353,25 @@ def _determine_project(so_line):
                         project = map_so_project[so_line.order_id.id]
                 if not so_line.task_id:
                     so_line._timesheet_create_task(project=project)
+
+    def _prepare_invoice_line(self, **optional_values):
+        """
+            If the sale order line isn't linked to a sale order which already have a default analytic account,
+            this method allows to retrieve the analytic account which is linked to project or task directly linked
+            to this sale order line, or the analytic account of the project which uses this sale order line, if it exists.
+        """
+        values = super(SaleOrderLine, self)._prepare_invoice_line(**optional_values)
+        if values['analytic_account_id']:
+            pass
+        elif self.project_id.analytic_account_id:
+            values['analytic_account_id'] = self.project_id.analytic_account_id
+        elif self.task_id.project_id.analytic_account_id:
+            values['analytic_account_id'] = self.task_id.project_id.analytic_account_id
+        elif self.is_service and not self.is_expense:
+            # Maybe should I define a One2many field reflecting the inverse relations to project.project.sale_line_id and project.task.sale_line_id ?
+            project_ids = self.env['project.project'].search([
+                '|', ('sale_line_id', '=', self.id), ('task_ids.sale_line_id', '=', self.id)
+            ])
+            if len(project_ids.analytic_account_id) == 1:
+                values['analytic_account_id'] = project_ids.analytic_account_id
+        return values
```

Comment:
In this method, we should give the id/ids rather than the recordset in the `values` variable, shouldn't we?

Reply:
Yes you're right :)